data_model/src/metadata.rs

```diff
@@ -0,0 +1,324 @@
+                _ => return None,
+            };
+        }
+        map.remove(key)
+    }
+
+    /// Insert the given [`Value`] into the given path. If the path is
+    /// complete, check the limits and only then insert. The creation
+    /// of the path is the responsibility of the user.
+    ///
+    /// # Errors
+    /// - If the path is empty.
+    /// - If one of the intermediate keys is absent.
+    /// - If some intermediate key is a leaf node.
+    pub fn nested_insert_with_limits(
+        &mut self,
+        path: &Path,
+        value: Value,
+        limits: Limits,
+    ) -> Result<Option<Value>> {
+        if self.map.len() >= limits.max_len as usize {
+            return Err(eyre!(
+                "Metadata length limit is reached: {}",
+                limits.max_len
+            ));
+        }
+        let key = path.iter().last().ok_or_else(|| eyre!("Empty path"))?;
+        let mut layer = self;
+        for k in path.iter().take(path.len() - 1) {
+            layer = match layer
+                .map
+                .get_mut(k)
+                .ok_or_else(|| eyre!("No metadata for key {} in path. Path is malformed.", k))?
+            {
+                Value::LimitedMetadata(data) => data,
+                _ => return Err(eyre!("Path contains non-metadata segments at key {}.", k)),
+            };
+        }
+        check_size_limits(key, &value, limits)?;
+        layer.insert_with_limits(key.to_string(), value, limits)
+    }
+
+    /// Insert [`Value`] under the given key.  Returns `Some(value)`
+    /// if the value was already present, `None` otherwise.
+    ///
+    /// # Errors
+    /// Fails if `max_entry_byte_size` or `max_len` from `limits` are exceeded.
+    pub fn insert_with_limits(
+        &mut self,
+        key: Name,
+        value: Value,
+        limits: Limits,
+    ) -> Result<Option<Value>> {
+        if self.map.len() >= limits.max_len as usize && !self.map.contains_key(&key) {
+            return Err(eyre!(
+                "Metadata length limit is reached: {}",
+                limits.max_len
+            ));
+        }
+        check_size_limits(&key, &value, limits)?;
+        Ok(self.map.insert(key, value))
+    }
+
+    /// Returns a `Some(reference)` to the value corresponding to
+    /// the key, and `None` if not found.
+    #[inline]
+    pub fn get<K: Ord + ?Sized>(&self, key: &K) -> Option<&Value>
+    where
+        Name: Borrow<K>,
+    {
+        self.map.get(key)
+    }
+
+    /// Removes a key from the map, returning the owned
+    /// `Some(value)` at the key if the key was previously in the
+    /// map, else `None`.
+    #[inline]
+    pub fn remove<K: Ord + ?Sized>(&mut self, key: &K) -> Option<Value>
+    where
+        Name: Borrow<K>,
+    {
+        self.map.remove(key)
+    }
+}
+
+fn check_size_limits(key: &str, value: &Value, limits: Limits) -> Result<()> {
+    let entry_bytes: Vec<u8> = (key, value.clone()).encode();
```

Comment:
nitpick maybe, but I think I'd rather clone before calling a function then inside it, just so that the caller is aware of the added cost